File: libs/shared-models/src/entity.ts
import {
  BoundingBox, Classification, Coordinate, DamageAssessment,
  OperationalStatus, PaginationRequest, PaginationResponse,
  TrackEnvironment, TrackProcessingState,
} from './common';
import { KinematicState, Orientation } from './kinematics';
import { PlatformData } from './platform-data';

// Canonical definitions in @sentinel/proto-gen (proto/entity.proto)
export enum EntityType {
  UNKNOWN = 'UNKNOWN',
  PERSON = 'PERSON',
  VEHICLE = 'VEHICLE',
  VESSEL = 'VESSEL',
  AIRCRAFT = 'AIRCRAFT',
  FACILITY = 'FACILITY',
  EQUIPMENT = 'EQUIPMENT',
  UNIT = 'UNIT',
  SIGNAL = 'SIGNAL',
  CYBER = 'CYBER',
  SENSOR = 'SENSOR',
  SATELLITE = 'SATELLITE',
  DRONE = 'DRONE',
}

export enum EntitySource {
  HUMINT = 'HUMINT',
  SIGINT = 'SIGINT',
  GEOINT = 'GEOINT',
  OSINT = 'OSINT',
  MASINT = 'MASINT',
  CYBER = 'CYBER',
  MANUAL = 'MANUAL',
  AIS = 'AIS',
  ADS_B = 'ADS_B',
  LINK16 = 'LINK16',
  GPS = 'GPS',
  RADAR = 'RADAR',
  CELESTRAK = 'CELESTRAK',
  OPENSKY = 'OPENSKY',
  ADSB_LOL = 'ADSB_LOL',
}

export enum Affiliation {
  FRIENDLY = 'FRIENDLY',
  HOSTILE = 'HOSTILE',
  NEUTRAL = 'NEUTRAL',
  UNKNOWN = 'UNKNOWN',
  ASSUMED_FRIENDLY = 'ASSUMED_FRIENDLY',
  SUSPECT = 'SUSPECT',
  PENDING = 'PENDING',
}

export enum CharacterizationState {
  ASSESSED = 'ASSESSED',
  ASSUMED = 'ASSUMED',
  SUSPECTED = 'SUSPECTED',
  UNCHARACTERIZED = 'UNCHARACTERIZED',
}

export interface Entity {
  id: string;
  entityType: EntityType;
  name: string;
  description?: string;
  source: EntitySource;
  classification: Classification;
  feedId?: string;
  position?: Coordinate;
  heading?: number;
  speedKnots?: number;
  course?: number;
  milStd2525dSymbol?: string;
  metadata: Record<string, string>;
  affiliations: string[];
  createdAt: string;
  updatedAt: string;
  lastSeenAt?: string;

  // UC2-informed identity & confidence
  affiliation?: Affiliation;
  identityConfidence?: number;
  characterization?: CharacterizationState;

  // Track context
  trackEnvironment?: TrackEnvironment;
  trackProcessingState?: TrackProcessingState;

  // Extended kinematics
  altitude?: number;
  orientation?: Orientation;
  kinematics?: KinematicState;

  // Operational status
  operationalStatus?: OperationalStatus;
  damageAssessment?: DamageAssessment;
  damageConfidence?: number;

  // Physical characteristics
  dimensions?: {
    length?: number;
    width?: number;
    height?: number;
  };
  countryOfOrigin?: string;

  // Protocol-specific structured data
  platformData?: PlatformData;
  sourceEntityId?: string;

  // Measurement quality
  circularError?: number;
  lastObservationSource?: string;
}

export enum EntityEventType {
  CREATED = 'CREATED',
  UPDATED = 'UPDATED',
  DELETED = 'DELETED',
  POSITION_UPDATE = 'POSITION_UPDATE',
}

export interface EntityEvent {
  eventId: string;
  entity: Entity;
  eventType: EntityEventType;
  timestamp: string;
}

export interface CreateEntityRequest {
  entityType: EntityType;
  name: string;
  description?: string;
  source: EntitySource;
  classification?: Classification;
  feedId?: string;
  position?: Coordinate;
  heading?: number;
  speedKnots?: number;
  course?: number;
  milStd2525dSymbol?: string;
  metadata?: Record<string, string>;
  affiliations?: string[];

  // UC2-informed fields
  affiliation?: Affiliation;
  identityConfidence?: number;
  characterization?: CharacterizationState;
  trackEnvironment?: TrackEnvironment;
  trackProcessingState?: TrackProcessingState;
  altitude?: number;
  orientation?: Orientation;
  kinematics?: KinematicState;
  operationalStatus?: OperationalStatus;
  damageAssessment?: DamageAssessment;
  damageConfidence?: number;
  dimensions?: { length?: number; width?: number; height?: number };
  countryOfOrigin?: string;
  platformData?: PlatformData;
  sourceEntityId?: string;
  circularError?: number;
  lastObservationSource?: string;
}

export interface UpdateEntityRequest {
  name?: string;
  description?: string;
  source?: EntitySource;
  classification?: Classification;
  position?: Coordinate;
  heading?: number;
  speedKnots?: number;
  course?: number;
  milStd2525dSymbol?: string;
  metadata?: Record<string, string>;
  affiliations?: string[];

  // UC2-informed fields
  affiliation?: Affiliation;
  identityConfidence?: number;
  characterization?: CharacterizationState;
  trackEnvironment?: TrackEnvironment;
  trackProcessingState?: TrackProcessingState;
  altitude?: number;
  orientation?: Orientation;
  kinematics?: KinematicState;
  operationalStatus?: OperationalStatus;
  damageAssessment?: DamageAssessment;
  damageConfidence?: number;
  dimensions?: { length?: number; width?: number; height?: number };
  countryOfOrigin?: string;
  platformData?: PlatformData;
  sourceEntityId?: string;
  circularError?: number;
  lastObservationSource?: string;
}

export interface QueryEntitiesRequest extends PaginationRequest {
  boundingBox?: BoundingBox;
  entityTypes?: EntityType[];
  sources?: EntitySource[];
  classificationMax?: Classification;
  textQuery?: string;
}

export interface QueryEntitiesResponse {
  entities: Entity[];
  pagination: PaginationResponse;
}

File: libs/common-go/models/entity.go
package models

// Entity represents the full entity model matching the database schema.
type Entity struct {
	ID                     string                 `json:"id"`
	EntityType             string                 `json:"entityType"`
	Name                   string                 `json:"name"`
	Description            *string                `json:"description,omitempty"`
	Source                 string                 `json:"source"`
	Classification         string                 `json:"classification"`
	FeedID                 *string                `json:"feedId,omitempty"`
	Position               *Coordinate            `json:"position,omitempty"`
	Heading                *float64               `json:"heading,omitempty"`
	SpeedKnots             *float64               `json:"speedKnots,omitempty"`
	Course                 *float64               `json:"course,omitempty"`
	MilStd2525dSymbol      *string                `json:"milStd2525dSymbol,omitempty"`
	Metadata               map[string]interface{} `json:"metadata"`
	Affiliations           []string               `json:"affiliations"`
	CreatedAt              string                 `json:"createdAt"`
	UpdatedAt              string                 `json:"updatedAt"`
	LastSeenAt             *string                `json:"lastSeenAt,omitempty"`
	Affiliation            *string                `json:"affiliation,omitempty"`
	IdentityConfidence     *int                   `json:"identityConfidence,omitempty"`
	Characterization       *string                `json:"characterization,omitempty"`
	TrackEnvironment       *string                `json:"trackEnvironment,omitempty"`
	TrackProcessingState   *string                `json:"trackProcessingState,omitempty"`
	Altitude               *float64               `json:"altitude,omitempty"`
	Orientation            *Orientation           `json:"orientation,omitempty"`
	Kinematics             *KinematicState         `json:"kinematics,omitempty"`
	OperationalStatus      *string                `json:"operationalStatus,omitempty"`
	DamageAssessment       *string                `json:"damageAssessment,omitempty"`
	DamageConfidence       *int                   `json:"damageConfidence,omitempty"`
	DimensionLength        *float64               `json:"dimensionLength,omitempty"`
	DimensionWidth         *float64               `json:"dimensionWidth,omitempty"`
	DimensionHeight        *float64               `json:"dimensionHeight,omitempty"`
	CountryOfOrigin        *string                `json:"countryOfOrigin,omitempty"`
	PlatformData           *PlatformData          `json:"platformData,omitempty"`
	SourceEntityID         *string                `json:"sourceEntityId,omitempty"`
	CircularError          *float64               `json:"circularError,omitempty"`
	LastObservationSource  *string                `json:"lastObservationSource,omitempty"`
	AgeoutState            string                 `json:"ageoutState,omitempty"`
	Deleted                bool                   `json:"deleted,omitempty"`
}
